### PR TITLE
Adapt NodeIPC module from cardano-sl

### PR DESCRIPTION
**Pull request by rvl**
Relates to issue #144 

# Overview

- Takes the cardano-sl/cardano-shell NodeIPC code, cleans it up, and splits the general NodeJS child_process IPC protocol out from the Daedalus/Cardano specific protocol, improves the exception handling, changes the logging, and just makes it simpler.
- Integrates DaedalusIPC into the server and launcher.
- Adds an integration test which starts the launcher from nodejs, sending `QueryPort`, waiting for `ReplyPort`, as Daedalus does.
- Adds a test for launcher `--state-dir` in the same Spec.

### CODE REVIEW

**Comment by KtorZ** [comment]
~~@rvl I am afraid you've done some double-work done on this PR :s ... ~~

~~Have you looked into `cardano-shell` (as mentioned in #144) -> https://github.com/input-output-hk/cardano-shell/tree/master/src/Cardano/Shell/NodeIPC ?~~

Nevermind, just saw your comment on your development plan:

> it's not suitable for re-use because testing code is mixed up with message protocol code which is mixed up with the ipc code.

File: lib/core/cardano-wallet-core.cabal [hunk 1]
```diff
@@ -76,6 +78,7 @@ library
       Cardano.Wallet.Api
       Cardano.Wallet.Api.Server
       Cardano.Wallet.Api.Types
+      Cardano.Wallet.DaedalusIPC
```
Comment (KtorZ) [inline lib/core/cardano-wallet-core.cabal:81]: Not a big fan of referring to `Daedalus` here. The IPC stuff can be seen as a low-level protocol to allow a client to perform some process management task of the wallet server. Of course, the primary user is Daedalus, but I don't want us to fall in the trap of "making stuff just for the frontend" as a `Daedalus***` module would suggest. This already caused us quite some pain with the old API that has some non-sensical stuff into it. 

In short: let's build more generic interfaces that serve a well-defined purpose.

File: lib/core/src/Cardano/Wallet/DaedalusIPC.hs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+daedalusIPC :: Int -> IO ()
+daedalusIPC port = withNodeChannel (pure . msg) action >>= \case
+    Right act -> do
+        sayErr "[INFO] Daedalus IPC server starting"
+        act
+    Left NodeChannelDisabled -> do
+        sayErr "[INFO] Daedalus IPC is not enabled"
+        sleep
+    Left (NodeChannelBadFD err) ->
+        sayErr $ "[ERROR] Starting Daedalus IPC: " <> err
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/DaedalusIPC.hs]: Now that we've the logging infrastructure setup, I'd suggest to switch to using the trace object instead of any other outputting method :+1:

File: lib/http-bridge/test/integration/js/mock-daedalus.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        console.log("JS: response from wallet: " + res.statusCode);
+        res.resume();
+        res.on("end", () => {
+          console.log("JS: request response from wallet finished, exiting.");
+          process.exit(0);
+        });
+      });
+    }
+  });
+}
```
Comment (KtorZ) [inline lib/http-bridge/test/integration/js/mock-daedalus.js:63]: Would be worth adding a negative test case with Daedalus sending some gibberish at first, to see that the `ParseFailure` message gets triggered correctly :+1:

File: lib/core/test/integration/Test/Integration/Scenario/CLI/Server.hs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+--
+-- cardano-wallet-launcher depends on cardano-wallet-http-bridge so that it can
+-- import the HttpBridge module.
+--
+-- This package (cardano-wallet-http-bridge) should have
+-- build-tool-depends: cardano-wallet:cardano-wallet-launcher so that it can
+-- run launcher in the tests. But that dependency can't be expressed in the
+-- cabal file, because otherwise there would be a cycle.
+--
+-- So one hacky way to work around it is by running programs under "stack exec".
```
Comment (KtorZ) [inline lib/core/test/integration/Test/Integration/Scenario/CLI/Server.hs:69]: :+1:

File: lib/core/src/Cardano/Wallet/DaedalusIPC.hs [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+ipcListener
+    :: forall msgin msgout. (FromJSON msgin, ToJSON msgout)
+    => Handle
+    -> (Either Text msgin -> IO (Maybe msgout))
+    -> MVar msgout
+    -> IO ()
+ipcListener handle onMsg chan = do
+    hSetNewlineMode handle noNewlineTranslation
+    catch (race_ replyLoop sendLoop) onIOError
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/DaedalusIPC.hs]: Nice pattern. I'd have a slight preference for `concurrently` instead of `race_` since we aren't actually expecting any of the two sides to eventually terminate (which `race_` conveys). We still get the cancellation behavior with `concurrently` if I am correct.

File: lib/core/src/Cardano/Wallet/DaedalusIPC.hs [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    sendMsg :: msgout -> IO ()
+    sendMsg = sendMessage handle . encode
+
+    maybeSend :: Maybe msgout -> IO ()
+    maybeSend = maybe (pure ()) (putMVar chan)
+
+    onIOError :: IOError -> IO ()
+    onIOError err = do
+      sayErrString $ "[ERROR] Exception caught in DaedalusIPC: " <> show err
+      when (isEOFError err) $ sayErr "[DEBUG] it's an eof"
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/DaedalusIPC.hs]: An `EOF` here would mean that Daedalus / the client has closed the connection on their hand, right ?

File: lib/core/cardano-wallet-core.cabal (hunk 1 above)
Comment (KtorZ) [reply lib/core/cardano-wallet-core.cabal:81]: I agree that `NodeIPC` was rather confusing, so maybe have something like `NodeJSIPC` or `JSClientIPC`

File: lib/core/src/Cardano/Wallet/DaedalusIPC.hs [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+    toEncoding = genericToEncoding aesonOpts
+
+daedalusIPC :: Int -> IO ()
+daedalusIPC port = withNodeChannel (pure . msg) action >>= \case
+    Right act -> do
+        sayErr "[INFO] Daedalus IPC server starting"
+        act
+    Left NodeChannelDisabled -> do
+        sayErr "[INFO] Daedalus IPC is not enabled"
+        sleep
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/DaedalusIPC.hs:120]: Why sleeping here? What would be the consequence of returning `()` ? Do we expect `daedalusIPC` to never return? If so, I'd rather have it mentioned in a function comment to explain this behavior.

File: exe/wallet/Main.hs [hunk 8]
```diff
@@ -385,9 +389,11 @@ execHttpBridge args _ = do
     waitForConnection nw defaultRetryPolicy
     let tl = HttpBridge.newTransactionLayer @n
     wallet <- newWalletLayer @_ @(HttpBridge n) tracer block0 db nw tl
-    let logStartup port = TIO.hPutStrLn stderr $
+    let logStartup port = logInfo tracer $
             "Wallet backend server listening on: " <> toText port
-    Server.start logStartup walletPort wallet
+    Server.withListeningSocket walletPort $ \(port, socket) -> do
+        let settings = Server.mkWarpSettings logStartup port
+        race_ (daedalusIPC port) (Server.startOnSocket settings socket wallet)
```
Comment (KtorZ) [inline exe/wallet/Main.hs]: Ah! I supposed this is why we make the `daedalusIPC` hangs forever. I believe that switching for `concurrently` instead of `race_` would remove that limitation since the action could return without cancelling the `startOnSocket`.

**Review by KtorZ — APPROVED** [review]
LGTM overall, but haven't really got time to play with it.


Also, some modification regarding ports have already been included in #407

File: lib/core/cardano-wallet-core.cabal (hunk 1 above)
Comment (rvl) [reply lib/core/cardano-wallet-core.cabal:81]: One of our main requirements is working with the Daedalus frontend. This ipc protocol is unique to Daedalus. I'm not aware of any requirements for integrating with some imaginary software that doesn't exist yet, but we can change the name if that happens.

It would be better to document the ways of starting cardano-wallet server on a port, so that prospective users can assess whether any of the schemes would work for them.

File: lib/core/src/Cardano/Wallet/DaedalusIPC.hs (hunk 2 above)
Comment (rvl) [reply lib/core/src/Cardano/Wallet/DaedalusIPC.hs]: OK

File: lib/http-bridge/test/integration/js/mock-daedalus.js (hunk 3 above)
Comment (rvl) [reply lib/http-bridge/test/integration/js/mock-daedalus.js:63]: Added a test. You might have noticed the tests are a bit thin. That's because
1. cardano-shell has more tests, and we will probably move the code there
2. I think it's worth considering inheriting the socket from the parent process, rather than implementing some hokey message protocol.

File: lib/core/src/Cardano/Wallet/DaedalusIPC.hs (hunk 5 above)
Comment (rvl) [reply lib/core/src/Cardano/Wallet/DaedalusIPC.hs]: It really doesn't matter which one. Both of them imply the wrong thing. But if `concurrently` makes you happy, then I'll change it.

File: lib/core/src/Cardano/Wallet/DaedalusIPC.hs (hunk 6 above)
Comment (rvl) [reply lib/core/src/Cardano/Wallet/DaedalusIPC.hs]: Usually the exception is usually broken pipe. I think this was debug code that got left in cardano-sl from the DEVOPS-1234 investigation. I have removed the log anyway.

File: lib/core/src/Cardano/Wallet/DaedalusIPC.hs (hunk 7 above)
Comment (rvl) [reply lib/core/src/Cardano/Wallet/DaedalusIPC.hs:120]: Yes, it should have been commented. This is the error handling I want:
1. If set up is unsuccessful, return immediately, causing the cardano-wallet process to exit. Daedalus does not want zombie cardano-wallet processes hanging around that it can't communicate with.
2. If set up is successful, never return until Daedalus exits.
3. If not running within a nodejs channel, never return (simplifies calling code)

File: exe/wallet/Main.hs (hunk 8 above)
Comment (rvl) [reply exe/wallet/Main.hs]: This needs to be `race` so that the program exits when there is an error from `daedalusIPC`.

File: lib/core/src/Cardano/Wallet/DaedalusIPC.hs (hunk 7 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/DaedalusIPC.hs:120]: Okay, so we really assume that any termination is an error, hence the sleep when there's no node.js channel (not an error so to speak).

File: exe/wallet/Main.hs (hunk 8 above)
Comment (KtorZ) [reply exe/wallet/Main.hs]: Got you, [this comment](https://github.com/input-output-hk/cardano-wallet/pull/388#discussion_r293646245) of yours clarifies the choice of `race` indeed. I'd suggest to have the behavior documented either on the call-site or, somewhere in the DaedalusIPC. Maybe in both places.

**edit**: I see it's now the case

**Comment by KtorZ** [comment]
Merging into #414 to group PRs and avoid multiple merges because of CI issues due to stack major bump today ....
